homeassistant/components/recorder/auto_repairs/schema.py

```diff
@@ -0,0 +1,225 @@
+
+
+def get_precise_datetime() -> datetime:
+    """Get a datetime with a precise microsecond.
+
+    This time can't be accurately represented unless datetimes have µs precision
+    """
+    future_year = _get_future_year()
+    return datetime(future_year, 10, 6, microsecond=1, tzinfo=dt_util.UTC)
+
+
+def validate_table_schema_supports_utf8(
+    instance: Recorder,
+    table_object: type[DeclarativeBase],
+    columns: tuple[str, ...],
+    session_maker: Callable[[], Session],
+) -> set[str]:
+    """Do some basic checks for common schema errors caused by manual migration."""
+    schema_errors: set[str] = set()
+
+    # Lack of full utf8 support is only an issue for MySQL / MariaDB
+    if instance.dialect_name != SupportedDialect.MYSQL:
+        return schema_errors
+
+    # Try inserting some data which needs utf8mb4 support
+    try:
+        # Mark the session as read_only to ensure that the test data is not committed
+        # to the database and we always rollback when the scope is exited
+        with session_scope(session=session_maker(), read_only=True) as session:
+            db_object = table_object(**{column: UTF8_NAME for column in columns})
+            table = table_object.__tablename__
+            session.add(db_object)
+            try:
+                session.flush()
+            except OperationalError as err:
+                if err.orig and err.orig.args[0] == MYSQL_ERR_INCORRECT_STRING_VALUE:
+                    _LOGGER.debug(
+                        "Database %s statistics_meta does not support 4-byte UTF-8",
+                        table,
+                    )
+                    schema_errors.add(f"{table}.4-byte UTF-8")
+                    session.rollback()
+                else:
+                    raise
+    except Exception as exc:  # pylint: disable=broad-except
+        _LOGGER.exception("Error when validating DB schema: %s", exc)
+    return schema_errors
+
+
+def validate_db_schema_precision(
+    instance: Recorder,
+    table_object: type[DeclarativeBase],
+    columns: tuple[str, ...],
+    session_maker: Callable[[], Session],
+) -> set[str]:
+    """Do some basic checks for common schema errors caused by manual migration."""
+    schema_errors: set[str] = set()
+    # Wrong precision is only an issue for MySQL / MariaDB / PostgreSQL
+    if instance.dialect_name not in (
+        SupportedDialect.MYSQL,
+        SupportedDialect.POSTGRESQL,
+    ):
+        return schema_errors
+
+    precise_time_ts = get_precise_datetime().timestamp()
+
+    try:
+        # Mark the session as read_only to ensure that the test data is not committed
+        # to the database and we always rollback when the scope is exited
+        with session_scope(session=session_maker(), read_only=True) as session:
+            db_object = table_object(**{column: precise_time_ts for column in columns})
+            table = table_object.__tablename__
+            session.add(db_object)
+            session.flush()
```

Comment:
The try and flush need to be wrapped with an exception catch and rollback before raising

Reply:
Actually probably not since it means the db is broken